README.md
# popover-confirm
基于 Bootstrap，依赖 jQuery 的二次确认弱弹框

# Demo

使用 Node 开启本地模拟 server
```
node server e
```

打开浏览器，输入 demo 演示地址：

```
http://localhost:8000/demo/index.html
```

# Usage

```javascript
// ...

// 这个方法不写也行，可以在 show 方法里面合并
function showUpdateStatusConfirm( e, index, data ) {
    e.stopPropagation(); // 终止事件在传播过程的捕获、目标处理或起泡阶段进一步传播
    show( $(this), index, data);
}

// ...

function show( $trigger, index, data ) {
    var _this = this;
    var handleName = data.status === 1 ? '启用' : '禁用';

    PopoverConfirm.init({
        UID: data.id, // 数据中唯一标识符，比如 ID，UserID 等，以确保重复点击显示、隐藏不会闪烁
        title: '确定'+ handleName +'？',
        loadingContent: handleName + '生效中.........', // Popover 加载中提示文字
        $trigger: $trigger, // 触发者
        ajax: {
            config: { // ajax 配置
                type: 'post',
                dataType: 'json',
                contentType: 'application/json;charset=utf-8',
                url: URL.changeStatus,
                data: data,
            },
            callback: function( res ) { // success 回调函数
                if ( res.Status ) {
                    const miniMsg = new MiniMsg({
                        content: handleName + '成功',
                        duration: 1,
                        container: $('body'),
                        type: 'success'
                    });

                    PopoverConfirm.destroy(); // 销毁
                    miniMsg.animation(() => {
                        avalon.vmodels.main.data[ index ].Status = data.status; // 更改数据状态
                    });
                } else {
                    // 可以设置错误信息提示
                    PopoverConfirm.setContent({
                        title: '操作失败',
                        content: '<span class="text-danger">'+ res.Message +'</span>'
                    });
                }
            }
        }
    });
}

// ...
```

# [更新日志](https://github.com/ct-adc/popover-confirm/blob/master/CHANGELOG.md)

CHANGELOG.md
## 更新日志

### 1.2.1

版本列表后补

### 1.2.0

版本列表后补

### 1.1.0

版本列表后补

### 1.0.0

版本列表后补


server.js
var fs = require('fs');	// 用于处理本地文件
var http = require('http');	// Http服务器API
var server = new http.Server();	// 创建新的HTTP服务器

server.listen(8000);	// 监听端口8000

var httpOpt,
	debug = process.argv.splice(2)[0];

fs.readFile('./serverPackage.json', function(err, data) {
	if(err) { 
        throw err; 
	}

	var json = JSON.parse(data);

	httpOpt=json.config;

	start(json);
});

function httpGet(url, callback, jsonData) {
	if(debug){
		callback(jsonData);	

		return;
	}
	console.log(httpOpt.href + url);

	http.get(httpOpt.href + url, function(response) {
		var data = [];

		response.on('data', function(chunk) {
			data.push(chunk); 
		}); 

		response.on('end', function() { 
			data = (Buffer.concat(data)).toString();  
			console.log(data);
			callback(data);
			
		});
	});
}

function httpPost(url, request, callback, jsonData) {
	if(debug){
		callback(jsonData);

		return;
	}

	var postdata = "";

	request.addListener("data", function(postchunk) {
		postdata += postchunk;
	});

	request.addListener("end", function() {
		return postDeal(url,callback,postdata);
	});	
}

function jsonOut(response, data) {
	response.writeHead(200, {'Content-type':'application/json; charset=UTF-8'});
	response.write(data);
	response.end();
}

function getType(endTag) {
    var type=null;

    switch(endTag) {
        case 'html' :
        case 'htm' :
            type = 'text/html; charset=UTF-8';
            break;
        case 'js' :
            type = 'application/javascript; charset="UTF-8"';
            break;
        case 'css' :
            type = 'text/css; charset="UTF-8"';
            break;
        case 'txt' :
            type = 'text/plain; charset="UTF-8"';
            break;
        case 'manifest' :
            type = 'text/cache-manifest; charset="UTF-8"';
            break;
        default :
            type = 'application/octet-stream';
            break;
    }

    return type;
}

function postDeal(url, callback, postdata) { // 服务器代 POST 理请求
	var postData = postdata;

	var opt = {
			host: httpOpt.host,
			port: httpOpt.port,
			path: url,
			method: 'POST',
			headers: {  
	            "Content-Type": 'application/x-www-form-urlencoded',  
	            "Content-Length": postData.length  
	        } 
		},
		data = [],
		req = http.request(opt, function(response) {
			response.on('data', function(chunk) { 
				data.push(chunk); 
			});

			response.on('end', function() { 
				data = (Buffer.concat(data)).toString();     

				callback(data);			
			});
		});

	req.on('error', function(e) {
		console.log('problem with request: ' + e.message);
	});

	req.write(postData);
	req.end();
}

function getUrlObject(url, json) {
	var i,
		r = false,
		object = json.dataList;

	for(i = 0; i < object.length; i++) {
		if(object[i].url == url){
			r = object[i];
			break;	
		}
	}
	return r;
}

function start(json) {
	// 使用on方法注册时间处理
	server.on('request', function(request, response) { 
		var jsonData,
			urlObject,
			url = require('url').parse(request.url);

		console.log('req:\t' + url.path + '\t' + (new Date())); // log request

		if(url.pathname == '' || url.pathname == '/') {
			fs.readFile('./index.html', function(err, content) {
				if(err) {
					response.writeHead(404, { 'Content-Type':'text/plain; charset="UTF-8"' });
					response.write(err.message);
					response.end();
				} else {
					response.writeHead(200, { 'Content-Type':'text/html; charset=UTF-8' });
					response.write(content);
					response.end();
				}
			});

			return;
		}

		urlObject = getUrlObject(url.pathname, json);

		if(urlObject) {
			jsonData = JSON.stringify(urlObject.data);

			if(urlObject.type == "post") {	// post	request
				httpPost(url.path, request, function(data) {			
					jsonOut(response, data);
				}, jsonData);
			} else {	// get request
				httpGet(url.path, function(data) {				
					jsonOut(response,data);
				}, jsonData);
			}

			return;
		}

		var filename = url.pathname.substring(1); // request fileName
		var type = getType( filename.substring(filename.lastIndexOf('.')+1) ); // .fileType

		fs.readFile(filename, function(err, content) {
			if(err) {
				response.writeHead(404, { 'Content-Type':'text/plain; charset="UTF-8"' });
				response.write(err.message);
				response.end();
			} else {
				response.writeHead(200, { 'Content-Type':type });
				response.write(content);
				response.end();
			}
		});
	});
}

popover-confirm.js
/*!
 * v1.2.1
 * 
 * Copyright 2016 Live
 * Licensed MIT
 * 
 * Date: 2017-03-28
 */

define(function() {
	'use strict';

	var _options = {
			UID: '', // 唯一识别符，比如 ID，UserID 等，以确保重复点击显示、隐藏不会闪烁
			title: '提示', // 标题
			loadingContent: '努力加载中...', // ajax 加载中提示语
			ajax: { // ajax 配置项
				config: {
		            type: 'GET', // ajax 请求类型
		            url: '', // ajax 请求地址
		            data: {} // ajax 参数
		            // ... other
		        },
		        callback: function( res ) {} // ajax success 回调函数
			},
			$trigger: null // popover 触发者
		},
		_popoverSelector = '', // 当前弹出的 popover 的选择器
		_timestamp = (new Date()).getTime(); // 使用时间戳作为 确定、取消按钮 的 id 名称后缀
		
	var CONFIRM_TPL = '<div class="btn-group">' +
				 '<button type="button" id="' + _getIDName( 'J_Confirm' ) + '" class="btn btn-xs btn-primary"><i class="glyphicon glyphicon-ok-sign">&ensp;</i>确定</button>' +
				 '<button type="button" id="' + _getIDName( 'J_Cancel' ) + '" class="btn btn-xs btn-default"><i class="glyphicon glyphicon-remove-sign">&ensp;</i>取消</button>' +
			  '</div>';


	/**
     * 切换按钮的禁用属性
     */
    function _toggleBtnDisableAttr( $btn ) {
		$btn.prop('disabled', function (_, val) { return ! val; });
    }

    /**
     * 生成页面唯一的 id
     * @param  {String} idName
     * @return {String}        idName + postfix
     */
	function _getIDName( idName ) {
		return idName + _timestamp;
	}

	// 需要在模块引用的时候会自动执行（保证执行一次）
	function _bind() {
		// 点击 popover 以外的区域，隐藏 popover
		$( document ).on('click', function() {
			destroy();
		});

		// 启用、禁用 Popover 取消、确定点击绑定
		$( document ).on('click', '#' + _getIDName( 'J_Confirm' ), function() {
			_confirm( $(this) );
		});
		$( document ).on('click', '#' + _getIDName( 'J_Cancel' ), function() {
			destroy();
		});
	}
	
	/**
	 * 确认操作
	 */
	function _confirm() {
		var ajaxConfig = { // ajax 配置项
				type: _options.ajax.type,
				url: _options.ajax.url,
				data: _options.ajax.data,
				cache: false,
				complete: function() {
					_toggleBtnDisableAttr( _options.$trigger );
				},
				success: function( res ) {
					_options.ajax.callback( res );
				},
				error: function( error ) {
					setContent({
						title: '<span class="text-danger">操作失败</span>',
						content: '<span class="text-danger">网络错误，请刷新后重试！</span>'
					});
				}
			};

		$.extend( ajaxConfig, _options.ajax.config );

		_toggleBtnDisableAttr( _options.$trigger );

		setContent({
			content: _options.loadingContent
		});

		$.ajax( ajaxConfig );
	}

	/**
	 * 获取 popover 的 selector
	 * @return { Selector }
	 */
	function _getSelector() {
		return '#' + _options.$trigger.attr( 'aria-describedby' );
	}


	/**
	 * 阻止点击 popover 冒泡事件
	 */
	function _stopPropagation( selector ) {
		$( document ).on('click', selector, function( e ) {
			e.stopPropagation();
		});
	}

	/**
	 * 初始化
	 * @param  {Object} options 配置项，详见开头 _options 的说明
	 */
	function init( options ) {
		var isNewUID = (_options.UID !== options.UID);

		destroy();

		if ( isNewUID ) { // 点击的如果不是同一个按钮，则切换显示、隐藏
			$.extend( _options, options );

			_options.$trigger.popover({
				html: true,
				placement: _options.placement || 'left',
				trigger: 'manual',
				title: _options.title,
				content: CONFIRM_TPL
			});

			show();
			_popoverSelector = _getSelector();
			_stopPropagation( _popoverSelector ); // 必须先 show popover 才能获取到 selector
		}
	}

	/**
	 * 显示
	 */
	function show() {
		if ( _options.$trigger !== null ) {
			_options.$trigger.popover( 'show' );
		}
	}

	/**
	 * 隐藏
	 */
	function hide() {
		if ( _options.$trigger !== null ) {
			_options.$trigger.popover( 'hide' );
		}
	}

	/**
	 * 销毁
	 */
	function destroy() {
		if ( _options.$trigger !== null ) {
			_options.UID = ''; // 清空 uid
			_options.$trigger.popover( 'destroy' ); // 销毁 popover
			_options.$trigger.attr( "data-content", '' ); // 清空 content，防止影响后面配置 content
			_options.$trigger.attr( "data-original-title", '' ); // 清空 title，防止影响后面配置 title
		}
	}

	/**
	 * 设置 popover 的内容
	 * @param {Object} options 设置的内容对象
	 * {
	 *     title: '这是标题',
	 *     content: '这是内容'
	 * }
	 */
	function setContent( options ) {
		var title = options.title,
			content = options.content;

		if ( typeof title !== 'undefined' ) {
			_options.$trigger.attr( "data-original-title", title );
		}

		if ( typeof content !== 'undefined' ) {
			_options.$trigger.attr( "data-content", content );
		}

		_options.$trigger.popover( 'show' );
	}

	_bind();
	return {
		init: init,
		show: show,
		hide: hide,
		destroy: destroy,
		setContent: setContent
	};
});

popover-confirm-umd.js
/*!
 * v1.2.1
 * 
 * Copyright 2016 Live
 * Licensed MIT
 * 
 * Date: 2017-03-28
 */


!function (root, factory) {
	if (typeof exports === 'object') {
		module.exports = factory();
	} else if (typeof define === 'function' && define.amd) {
		define(factory);
	} else {
		root.popoverConfirm = factory();
	}
}(this, function() {
	'use strict';

	var _options = {
			UID: '', // 唯一识别符，比如 ID，UserID 等，以确保重复点击显示、隐藏不会闪烁
			title: '提示', // 标题
			loadingContent: '努力加载中...', // ajax 加载中提示语
			ajax: { // ajax 配置项
				config: {
		            type: 'GET', // ajax 请求类型
		            url: '', // ajax 请求地址
		            data: {} // ajax 参数
		            // ... other
		        },
		        callback: function( res ) {} // ajax success 回调函数
			},
			$trigger: null // popover 触发者
		},
		_popoverSelector = '', // 当前弹出的 popover 的选择器
		_timestamp = (new Date()).getTime(); // 使用时间戳作为 确定、取消按钮 的 id 名称后缀
		
	var CONFIRM_TPL = '<div class="btn-group">' +
				 '<button type="button" id="' + _getIDName( 'J_Confirm' ) + '" class="btn btn-xs btn-primary"><i class="glyphicon glyphicon-ok-sign">&ensp;</i>确定</button>' +
				 '<button type="button" id="' + _getIDName( 'J_Cancel' ) + '" class="btn btn-xs btn-default"><i class="glyphicon glyphicon-remove-sign">&ensp;</i>取消</button>' +
			  '</div>';


	/**
     * 切换按钮的禁用属性
     */
    function _toggleBtnDisableAttr( $btn ) {
		$btn.prop('disabled', function (_, val) { return ! val; });
    }

    /**
     * 生成页面唯一的 id
     * @param  {String} idName
     * @return {String}        idName + postfix
     */
	function _getIDName( idName ) {
		return idName + _timestamp;
	}

	// 需要在模块引用的时候会自动执行（保证执行一次）
	function _bind() {
		// 点击 popover 以外的区域，隐藏 popover
		$( document ).on('click', function() {
			destroy();
		});

		// 启用、禁用 Popover 取消、确定点击绑定
		$( document ).on('click', '#' + _getIDName( 'J_Confirm' ), function() {
			_confirm( $(this) );
		});
		$( document ).on('click', '#' + _getIDName( 'J_Cancel' ), function() {
			destroy();
		});
	}
	
	/**
	 * 确认操作
	 */
	function _confirm() {
		var ajaxConfig = { // ajax 配置项
				type: _options.ajax.type,
				url: _options.ajax.url,
				data: _options.ajax.data,
				cache: false,
				complete: function() {
					_toggleBtnDisableAttr( _options.$trigger );
				},
				success: function( res ) {
					_options.ajax.callback( res );
				},
				error: function( error ) {
					setContent({
						title: '<span class="text-danger">操作失败</span>',
						content: '<span class="text-danger">网络错误，请刷新后重试！</span>'
					});
				}
			};

		$.extend( ajaxConfig, _options.ajax.config );

		_toggleBtnDisableAttr( _options.$trigger );

		setContent({
			content: _options.loadingContent
		});

		$.ajax( ajaxConfig );
	}

	/**
	 * 获取 popover 的 selector
	 * @return { Selector }
	 */
	function _getSelector() {
		return '#' + _options.$trigger.attr( 'aria-describedby' );
	}


	/**
	 * 阻止点击 popover 冒泡事件
	 */
	function _stopPropagation( selector ) {
		$( document ).on('click', selector, function( e ) {
			e.stopPropagation();
		});
	}

	/**
	 * 初始化
	 * @param  {Object} options 配置项，详见开头 _options 的说明
	 */
	function init( options ) {
		var isNewUID = (_options.UID !== options.UID);

		destroy();

		if ( isNewUID ) { // 点击的如果不是同一个按钮，则切换显示、隐藏
			$.extend( _options, options );

			_options.$trigger.popover({
				html: true,
				placement: _options.placement || 'left',
				trigger: 'manual',
				title: _options.title,
				content: CONFIRM_TPL
			});

			show();
			_popoverSelector = _getSelector();
			_stopPropagation( _popoverSelector ); // 必须先 show popover 才能获取到 selector
		}
	}

	/**
	 * 显示
	 */
	function show() {
		if ( _options.$trigger !== null ) {
			_options.$trigger.popover( 'show' );
		}
	}

	/**
	 * 隐藏
	 */
	function hide() {
		if ( _options.$trigger !== null ) {
			_options.$trigger.popover( 'hide' );
		}
	}

	/**
	 * 销毁
	 */
	function destroy() {
		if ( _options.$trigger !== null ) {
			_options.UID = ''; // 清空 uid
			_options.$trigger.popover( 'destroy' ); // 销毁 popover
			_options.$trigger.attr( "data-content", '' ); // 清空 content，防止影响后面配置 content
			_options.$trigger.attr( "data-original-title", '' ); // 清空 title，防止影响后面配置 title
		}
	}

	/**
	 * 设置 popover 的内容
	 * @param {Object} options 设置的内容对象
	 * {
	 *     title: '这是标题',
	 *     content: '这是内容'
	 * }
	 */
	function setContent( options ) {
		var title = options.title,
			content = options.content;

		if ( typeof title !== 'undefined' ) {
			_options.$trigger.attr( "data-original-title", title );
		}

		if ( typeof content !== 'undefined' ) {
			_options.$trigger.attr( "data-content", content );
		}

		_options.$trigger.popover( 'show' );
	}

	_bind();
	return {
		init: init,
		show: show,
		hide: hide,
		destroy: destroy,
		setContent: setContent
	};
});